scripts/mgit

```diff
@@ -5,41 +5,118 @@
 # disruptive action, like power off or SD card extraction.
 # ----------------------------------------------------------------------------
 
-MARKER_BASE=$HOME/git_in_progress
+MGIT_DIR=$HOME/.mgit
+MGIT_REPOS=$MGIT_DIR/repos
+LOCK_TIMEOUT=1800
 
 function marked_git {
-    marker_dir=$MARKER_BASE/$1
-    mkdir -p $marker_dir
-    echo $2 > $marker_dir/url
-    echo $3 > $marker_dir/branch
-    shift 3
-    git "$@" || return 1
-    rm -rf $marker_dir
+    toplevel=$(git rev-parse --show-toplevel) || return 1
+    repo_name=$(basename $toplevel)
+    mgit_repo=$MGIT_REPOS/$repo_name
+    lock_file=$MGIT_REPOS/$repo_name/lock
+    marker_file=$MGIT_REPOS/$repo_name/git-in-progress
+    mkdir -p $mgit_repo
+    (
+    flock -w $LOCK_TIMEOUT 9 || exit 1
+
+    # save repo info
+    git remote get-url origin > $mgit_repo/url
+    sync
+    git rev-parse --abbrev-ref HEAD > $mgit_repo/branch
+    sync
+    echo $toplevel > $mgit_repo/toplevel
+    sync
+
+    # lay down the marker file
+    date > $marker_file
+    sync
+
+    # execute actual git command
+    git "$@" || exit 1
```

Comment:
we should continue deleting the mark file even the git command is failed.

the purpose for mgit recover is to recover if user unplugs power during git operations, not to recover if git fails due to like network issue or github down..